### ISSUE TITLE
Allow type aliases to be passed to generics inside their definition if that generic param 

### ISSUE THREAD

**Issue by JakeTunaley** [Microsoft/TypeScript#27332]
**TypeScript Version:**  3.1.0-dev.20180922

**Search Terms:** generic alias circular

**Code**

```ts
interface Generic<T> {
    foo (): T;
}

type ConcreteAlias = {
    foo (): ConcreteAlias;
};
type GenericAlias = Generic<GenericAlias>;
```

**Expected behavior:**
`ConcreteAlias` and `GenericAlias` should both compile without error, as they are equivalent and type aliases can be used inside themselves for function return types.

**Actual behavior:**
`GenericAlias` has the following error:

```
Type alias 'GenericAlias' circularly references itself.
```

**Suggestion:**
Allow type aliases to pass themselves as arguments to generics if it can be determined that that generic param is only ever used in ways that don't throw errors re: circular references.

**Playground Link:**
http://www.typescriptlang.org/play/#src=interface%20Generic%3CT%3E%20%7B%0D%0A%20%20%20%20foo%20()%3A%20T%3B%0D%0A%7D%0D%0A%0D%0Atype%20ConcreteAlias%20%3D%20%7B%0D%0A%20%20%20%20foo%20()%3A%20ConcreteAlias%3B%0D%0A%7D%3B%0D%0Atype%20GenericAlias%20%3D%20Generic%3CGenericAlias%3E%3B

**Comment by JakeTunaley** [comment]
For now the workaround for stuff like this is the following:

```ts
interface GenericAlias extends Generic<GenericAlias> {}
```

**Comment by RyanCavanaugh** [comment]
See #6230
